### PR TITLE
I 295 develop work

### PR DESCRIPTION
**Pull request by clevengr**
### Description of what the PR does
  Updates a number of classes in support of partial work on #295.  Updated classes include:  Problem, ProblemsPane, EditProblemFrame, EditProblemPane, and EditProblemSandboxPane.   The result is (what I believe to be) a working GUI for supporting Memory Limits.  (Note however that this doesn't mean Memory Limits now work; that requires additional updates for invoking a sandbox shell.  This PR is strictly about (hopefully) completing the GUI support for specifying Memory Limits.)

### Issue which the PR fixes
None; it is a *partial* fix for Issue 295.

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 10.1, Eclipse Version: 2019-12 (4.14.0), Java 1.8_201.

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

- Start a server with a contest with at least one problem.
- Start an Admin.
- Select the "Problems" tab.
- Select a problem and click the "Edit" button.
- Select the "Sandbox" tab.
- Test the following:
  - ability to select either "No Sandbox" or "PC2 Sandbox"
  - *inability* to select "Custom Sandbox"
  - ability to enter an arbitrary memory limit value when the PC2 Sandbox button is selected
  - *inability* to enter a non-numeric memory limit value
  - entering a non-negative memory limit value and then pressing "Update" saves the memory limit value in the problem
    - the new memory limit value displays on the "Problems Pane" in the "Memory Limit" column
    - re-editing the same problem causes the new memory limit value to appear in the "Memory Limit" textbox on the edit screen
  - entering a negative memory limit value causes the memory limit to be set to zero
  - Pressing each of the "Information Question Mark" icons causes an appropriate and helpful explanatory message to be displayed.

### Additional Information
  The code in EditProblemSandboxPane includes declarations for two "debug variables":  *debugShowPanelBorders*, which is useful when debugging GUI layout issues, and *debugAllowSandboxSelectionOnWindows*.  This latter variable allows testing the GUI selection capabilities on a Windows system even though eventually we are going to disallow such selection on Windows (until we get a working Windows sandbox).    Currently, *debugShowPanelBorders* is set false while *debugAllowSandboxSelectionOnWindows* is set true (to allow reviewers to test the GUI).  The *debugAllowSandboxSelectionOnWindows* variable needs to be set false in a production distribution...

### CODE REVIEW

File: src/edu/csus/ecs/pc2/core/model/Problem.java [hunk 1]
```diff
@@ -1992,31 +2008,11 @@ public void setVivaInputValidatorHasBeenRun(boolean vivaInputValidatorHasBeenRun
         vivaSettings.setVivaInputValidatorHasBeenRun(vivaInputValidatorHasBeenRun);
     }
 
-    /**
-     * Returns the flag indicating whether this Problem has a runtime memory limit associated with it.
-     * 
-     * @return the flag indicating whether or not this Problem has been configured with a memory limit.
-     */
-    public boolean hasMemoryLimit() {
-        return hasMemoryLimit;
```
Comment (lane55) [inline src/edu/csus/ecs/pc2/core/model/Problem.java:2001]: Can we remove the field hasMemoryLimit and change this method just...
return memoryLimitMB  > 0;
?

File: src/edu/csus/ecs/pc2/core/model/Problem.java [hunk 2]
```diff
@@ -211,8 +214,6 @@
     private VivaInputValidatorSettings vivaSettings = null;
 
 
-
```
Comment (lane55) [inline src/edu/csus/ecs/pc2/core/model/Problem.java:214]: Unable to leave a comment near:   public SandboxType getSandboxType() {
For backwards compatibility and to avoid NPEs should change to:
    public SandboxType getSandboxType() {
        if (sandboxType == null) {
            sandboxType  = SandboxType.NONE;
        }
        return sandboxType;
    }

File: src/edu/csus/ecs/pc2/ui/EditProblemPane.java [hunk 3]
```diff
@@ -690,232 +700,258 @@ public void enableUpdateButton() {
...
+                    } else {
+                        updateToolTip += ", Output Validator";
+                    }
+                }
 
-                    // see if the choice of which output validator (if any) to use has changed;
-                    // enable the Update button and update the tooltip if so
-                    if (problem.getValidatorType() != changedProblem.getValidatorType()) {
+                // see if the PC2 Validator options have changed; enable the Update button and update the tooltip if so
+                if (problem.getPC2ValidatorSettings() != null && changedProblem.getPC2ValidatorSettings() != null) {
```
Comment (lane55) [inline src/edu/csus/ecs/pc2/ui/EditProblemPane.java:721]: Is this related to a bug fix?   Validator field are not in the scope of sandbox changes.

File: src/edu/csus/ecs/pc2/ui/EditProblemFrame.java [hunk 4]
```diff
@@ -30,7 +30,7 @@
 
     private IInternalController controller;
 
-    private EditProblemPane problemPane = null;
```
Comment (lane55) [inline src/edu/csus/ecs/pc2/ui/EditProblemFrame.java:33]: Confirm that the only change in this file was a rename of problemPane, thanks.

File: src/edu/csus/ecs/pc2/ui/EditProblemPane.java [hunk 5]
```diff
@@ -690,232 +700,258 @@ public void enableUpdateButton() {
...
                         }
+                    }
 
-                        // check for changes in the specified interface being used
-                        if (problemSettings.isUseClicsValidatorInterface() != changedProblemSettings.isUseClicsValidatorInterface()
-                                || problemSettings.isUsePC2ValidatorInterface() != changedProblemSettings.isUsePC2ValidatorInterface()) {
+                    // check for changes in the actual validator program file
+                    String changedProblemValidatorFileName = changedProblemSettings.getCustomValidatorProgramName();
+                    if (changedProblemValidatorFileName != null && changedProblemValidatorFileName.length() > 0) {
+                        // TODO: there could be a problem here; pdf.getValidatorFile() seems to sometimes return null even though the problem has a validator...
```
Comment (lane55) [inline src/edu/csus/ecs/pc2/ui/EditProblemPane.java:783]: for this TODO, is the code updated to avoid this potential bug ?

File: src/edu/csus/ecs/pc2/ui/EditProblemPane.java [hunk 6]
```diff
@@ -5492,8 +5615,20 @@ private void logWarning(String message, Exception e) {
...
-        frame.getContentPane().add(new EditProblemPane());
+
+        //build an EditProblemFrame containing an EditProblemPane containing an EditProblemSandboxPane
+        EditProblemFrame frame = new EditProblemFrame();
+        
+        //put a contest and controller in the frame (which also puts it in the panes)
+        IInternalContest contest;
+        IInternalController controller;
+
+        SampleContest sampleContest = new SampleContest();
```
Comment (lane55) [inline src/edu/csus/ecs/pc2/ui/EditProblemPane.java:5626]: BUILD FAILURE.   This cause a build failure, SampleContest is not part of the src/ tree it is part of the test/ tree.
Error is: symbol:   class SampleContest
[javac]   location: class EditProblemPane

**Review by lane55 — CHANGES_REQUESTED** [review]
Need to fix the build failure using SampleContst.
Discuss other smaller changes.

File: src/edu/csus/ecs/pc2/core/model/Problem.java (hunk 1 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:2001]: Sure, we COULD do that.  I tend to want to isolate functions with accessors.  What if we later decide we have a different definition of what it means to "have a memory limit"?   If we keep it surrounded by an accessor -- and all other code always uses the accessor (which I think I tried to ensure), we only have to change the accessor to make such a change.  I think it's worth the one-line accessor to provide this improved code structure.

File: src/edu/csus/ecs/pc2/core/model/Problem.java (hunk 2 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:214]: Agreed; fixed.

File: src/edu/csus/ecs/pc2/ui/EditProblemFrame.java (hunk 4 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/ui/EditProblemFrame.java:33]: Confirmed.

File: src/edu/csus/ecs/pc2/ui/EditProblemPane.java (hunk 5 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/ui/EditProblemPane.java:783]: I don't know.  This TODO was already in the code from long ago; I didn't add it and it has nothing to do with Memory Limits.

File: src/edu/csus/ecs/pc2/ui/EditProblemPane.java (hunk 6 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/ui/EditProblemPane.java:5626]: Hmmm.... I'm not clear what the correct fix for this is.   Can you suggest a fix?   Or shall I just remove the test function (main)?

File: src/edu/csus/ecs/pc2/core/model/Problem.java (hunk 1 above)
Comment (lane55) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:2001]: Since I did not suggest removing or changing the sig,   My suggestion was to change the code inside
the accessor which is supported by your comment.   Per the comment for Problem:40
 public static final int DEFAULT_MEMORY_LIMIT_MB = 0 ;   //zero memory limit = "none", i.e. the problem can use all available memory
that indicates to me that memoryLimitMB > 0; is the condition for hasMemoryLimit.

**Comment by lane55** [comment]
Suggested Fix for the SampleContest unresolved external/class in EditProblemPane.
Move the main method into the unit test class EditProblemPaneTest, main will be ignored
and not cause a build failure.

File: src/edu/csus/ecs/pc2/core/model/Problem.java (hunk 1 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/core/model/Problem.java:2001]: Ok, I misinterpreted your comment.  But:  there is something wrong.  The entire method `hasMemoryLimit()` was removed entirely.  You are commenting about code that was removed.

File: src/edu/csus/ecs/pc2/ui/EditProblemPane.java (hunk 3 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/ui/EditProblemPane.java:721]: No, there is no change (or intent to change) something related to validators.  The "code change" you are seeing is a result of the "diff" routine being unable to deal with a different change:  altering the scope of this test.  That change was necessary in order to place the code which tests for changes in the sandbox in the correct level of scope.

File: src/edu/csus/ecs/pc2/ui/EditProblemPane.java (hunk 6 above)
Comment (clevengr) [reply src/edu/csus/ecs/pc2/ui/EditProblemPane.java:5626]: I've commented out the offending code and will push an update.

**Review by lane55 — APPROVED** [review]
Approved.
